### PR TITLE
osdc: After write try merge bh.

### PR DESCRIPTION
**Pull request by majianpeng**
Fixes:http://tracker.ceph.com/issues/17270

Signed-off-by: Jianpeng Ma jianpeng.ma@intel.com

### CODE REVIEW

File: src/osdc/ObjectCacher.cc [hunk 1]
```diff
@@ -182,6 +182,7 @@ void ObjectCacher::Object::merge_left(BufferHead *left, BufferHead *right)
 
   // hose right
   delete right;
+  right = nullptr; //see bh_write_commit
```
Comment (liewegas) [inline src/osdc/ObjectCacher.cc]: Isn't this a local variable?

File: src/osdc/ObjectCacher.cc [hunk 2]
```diff
@@ -1156,6 +1159,13 @@ void ObjectCacher::bh_write_commit(int64_t poolid, sobject_t oid,
 		       << dendl;
       }
     }
+
+    while (!hit.empty()) {
+      BufferHead *bh = hit.front();
+      hit.pop_front();
+      if (bh)
```
Comment (liewegas) [inline src/osdc/ObjectCacher.cc]: bh should always be non-null?

Comment (majianpeng) [inline src/osdc/ObjectCacher.cc]: #11262 . In hit, the first may merged w/ the second. So i need know which be merged(delete). Yes as you said, set the local var to nullptr after delete don't indicate the related var in hit. But change the merge_left right to BufferHead **right. I think it's not a goog method.

**Comment by majianpeng** [comment]
@liewegas . I think this version can work.

**Comment by liewegas** [comment]
Can you remove the Fixes line from the commit?  It make me think this fixes a bug but it doesn't--it just optimizes in a non-buggy way :)  Thanks!

**Comment by jcsp** [comment]
Tested here: http://pulpito.ceph.com/jspray-2016-11-10_21:38:43-fs-wip-jcsp-testing-20161108-distro-basic-mira/

**Pull request merged**